### PR TITLE
Data Slice

### LINKED ISSUE #65: Add Context
User can add parameter in labeling function to access context.

```python
def labeling_function(df, context):
    context.window  # (start_time, stop_time)
    context.gap  # (start_time, stop_time)
    context.slice  # current slice
    ...
```

### PR DESCRIPTION
**Pull request by jeff-hernandez**
This PR creates a data slice class to add contextual attributes. Resolves #65.

```python
def labeling_function(df):
    df.window  # (start_time, stop_time)
    df.gap  # (start_time, stop_time)
    df.slice  # current slice number
    ...
```

### CODE REVIEW

File: composeml/label_maker.py [hunk 1]
```diff
@@ -213,12 +224,11 @@ def slice(self,
             minimum_data (str) : Minimum data before starting search. Default value is first time of index.
             gap (str or int) : Time between examples. Default value is window size.
                 If an integer, search will start on the first event after the minimum data.
-            metadata (bool) : Whether to return metadata about the data slice. Default value is False.
             drop_empty (bool) : Whether to drop empty slices. Default value is True.
             verbose (bool) : Whether to print metadata about slice. Default value is False.
 
         Returns:
-            DataFrame : Slice of data.
+            DataSlice : Retruns data slice.
```
Comment (kmax12) [inline composeml/label_maker.py]: typo in "Retruns data slice"

Comment (jeff-hernandez) [reply composeml/label_maker.py]: fixed typo

File: composeml/label_maker.py [hunk 2]
```diff
@@ -172,19 +202,19 @@ def iloc(index, i):
                     if df_slice.index.size > 1 and is_overlap.any():
                         df_slice = df_slice[~is_overlap]
 
-            metadata['window'] = (cutoff_time, window_end)
+            df_slice.context.window = (cutoff_time, window_end)
 
             if isinstance(gap, int):
                 gap_end = iloc(df.index, gap)
-                metadata['gap'] = (cutoff_time, gap_end)
+                df_slice.context.gap = (cutoff_time, gap_end)
```
Comment (kmax12) [inline composeml/label_maker.py]: is `cutoff_time` the right argument here? shouldn't the start of the gap be the end of the pervious slices window?

File: composeml/label_maker.py [hunk 3]
```diff
@@ -232,21 +254,19 @@ def slice(self,
...
 
-        for key, df in df.groupby(self.target_entity):
-            slices = self._get_slices(df=df, gap=gap, min_data=minimum_data, drop_empty=drop_empty)
+        for key, instance in df.groupby(self.target_entity):
+            slices = self._get_slices(df=instance, gap=gap, min_data=minimum_data, drop_empty=drop_empty)
 
-            for df_slice, df_metadata in slices:
-                df_metadata[self.target_entity] = key
+            for df in slices:
+                df.context.target_entity = self.target_entity
```
Comment (kmax12) [inline composeml/label_maker.py]: any way to move this context setting logic into `_get_slices`? I think it'd be better to have that function return the slices all ready to go

File: composeml/label_maker.py [hunk 4]
```diff
@@ -232,21 +254,19 @@ def slice(self,
...
+                df.context.target_instance = key
 
                 if verbose:
-                    self.print_slice(df_metadata)
-
-                if metadata:
-                    df_slice = df_slice, df_metadata
+                    self.print_slice(df)
```
Comment (kmax12) [inline composeml/label_maker.py]: what if `print_slice` was defined as the repr for `DataSlice` and you just do `print(df)` here?

Comment (jeff-hernandez) [reply composeml/label_maker.py]: I like that idea. Will override `__repr__` for `print(repr(df))`.

File: composeml/label_maker.py (hunk 3 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: Yes, will move in `_get_slices`.

File: composeml/label_maker.py (hunk 2 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: The start of the gap is based off the cutoff time to allow for rolling windows.

File: composeml/label_maker.py [hunk 5]
```diff
@@ -94,6 +94,35 @@ def to_offset(value):
...
+        Args:
+            gap (tuple) : Start and stop time for gap.
+            window (tuple) : Start and stop time for window.
+            slice (int) : Slice number.
+            target_entity (int) : Target entity.
+            target_instance (int) : Target instance.
+        """
+        self.gap = gap or (None, None)
+        self.window = window or (None, None)
+        self.slice = slice
```
[the remark this answers is not in the document]
Comment (jeff-hernandez) [reply composeml/label_maker.py]: updated to `slice_number`

**Pull request merged**